cpp/src/arrow/record_batch.cc

```diff
@@ -253,14 +253,65 @@ inline void ConvertColumnsToTensor(const RecordBatch& batch, uint8_t* out) {
   using CType = typename arrow::TypeTraits<DataType>::CType;
   auto* out_values = reinterpret_cast<CType*>(out);
 
-  // Loop through all of the columns
-  for (int i = 0; i < batch.num_columns(); ++i) {
-    const auto* in_values = batch.column(i)->data()->GetValues<CType>(1);
-
-    // Copy data of each column
-    memcpy(out_values, in_values, sizeof(CType) * batch.num_rows());
-    out_values += batch.num_rows();
-  }  // End loop through columns
+  if (TypeTraits<DataType>::type_singleton() ==
+      batch.column(0)->type()) {  // If all columns are of same data type
+    // Loop through all of the columns
+    for (int i = 0; i < batch.num_columns(); ++i) {
+      const auto& arr = *batch.column(i);
+      auto data = arr.data();
+      const auto& in_values = data->GetValues<CType>(1);
+
+      // Copy data of each column
+      memcpy(out_values, in_values, sizeof(CType) * batch.num_rows());
+      out_values += batch.num_rows();
+    }  // End loop through columns
+
+  } else {  // If columns have mixed data type
```

Comment:
We might want to keep the simpler memcpy for those columns where the column type matches the result type? 
Another way to organize the code is to first loop over the columns, and only then switch between memcpy or the loop, depending on the column type.

Reply:
That's a good idea! Will do